File: music_car/android/app/src/main/java/com/icar/musiccar/network/AIChoreographer.kt
/**
 * 音乐小车 — AI 编舞客户端（Android / Kotlin）
 *
 * 调用小车后端 /api/choreograph/* 接口，实现 AI 生成 + 微调编舞。
 *
 * 依赖: OkHttp + kotlinx.serialization
 */

package com.icar.musiccar.network

import com.icar.musiccar.AppConstants
import com.icar.musiccar.models.ChoreographRequest
import com.icar.musiccar.models.ChoreographResponse
import com.icar.musiccar.models.DanceRoutine
import kotlinx.serialization.json.Json
import kotlinx.serialization.encodeToString
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

class AIChoreographer {
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private var baseUrl: String = "http://${AppConstants.DEFAULT_CAR_IP}:${AppConstants.DEFAULT_CAR_PORT}"
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(AppConstants.LLM_TIMEOUT, TimeUnit.MILLISECONDS)
        .readTimeout(AppConstants.LLM_TIMEOUT, TimeUnit.MILLISECONDS)
        .build()

    fun setServer(ip: String, port: Int) {
        baseUrl = "http://$ip:$port"
    }

    /** AI 生成编舞 */
    suspend fun generate(req: ChoreographRequest): ChoreographResponse {
        return post("/api/choreograph/generate", req)
    }

    /** AI 微调已有编舞 */
    suspend fun refine(routine: DanceRoutine, feedback: String): ChoreographResponse {
        val data = mapOf("routine" to routine, "feedback" to feedback)
        return post("/api/choreograph/refine", data)
    }

    /** 校验编舞合法性（不调 LLM） */
    suspend fun validate(routine: DanceRoutine): ChoreographResponse {
        return post("/api/choreograph/validate", routine)
    }

    /** 查看当前 System Prompt */
    suspend fun getPrompt(): PromptResponse {
        return kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
            try {
                val request = Request.Builder()
                    .url("$baseUrl/api/choreograph/prompt")
                    .get()
                    .build()
                val response = client.newCall(request).execute()
                if (response.isSuccessful) {
                    val body = response.body?.string() ?: "{}"
                    json.decodeFromString<PromptResponse>(body)
                } else {
                    PromptResponse(ok = false)
                }
            } catch (_: Exception) {
                PromptResponse(ok = false)
            }
        }
    }

    // -- 内部方法 --

    private suspend fun post(path: String, data: Any): ChoreographResponse {
        val body = json.encodeToString(data).toRequestBody(jsonMediaType)
        val request = Request.Builder()
            .url("$baseUrl$path")
            .post(body)
            .build()

        return kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
            try {
                val response = client.newCall(request).execute()
                if (response.isSuccessful) {
                    val bodyStr = response.body?.string() ?: "{}"
                    json.decodeFromString<ChoreographResponse>(bodyStr)
                } else {
                    ChoreographResponse(ok = false, error = "HTTP ${response.code}")
                }
            } catch (e: IOException) {
                ChoreographResponse(ok = false, error = "网络错误: ${e.message}")
            }
        }
    }
}

// -- Prompt 响应 --

data class PromptResponse(
    val ok: Boolean,
    val prompt: String? = null,
    val backend: String? = null
)

File: music_car/ios/MusicCar/Network/AIChoreographer.swift
/**
 * 音乐小车 — AI 编舞客户端（iOS / Swift）
 *
 * 调用小车后端 /api/choreograph/* 接口，实现 AI 生成 + 微调编舞。
 */

import Foundation

class AIChoreographer {
    private var baseURL: String = "http://\(AppConstants.defaultCarIP):\(AppConstants.defaultCarPort)"
    private let session: URLSession
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()
    
    init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = AppConstants.llmTimeout
        config.timeoutIntervalForResource = AppConstants.llmTimeout
        session = URLSession(configuration: config)
    }
    
    func setServer(ip: String, port: Int) {
        baseURL = "http://\(ip):\(port)"
    }
    
    /// AI 生成编舞
    func generate(_ req: ChoreographRequest) async -> ChoreographResponse {
        return await post(path: "/api/choreograph/generate", body: req)
    }
    
    /// AI 微调已有编舞
    func refine(routine: DanceRoutine, feedback: String) async -> ChoreographResponse {
        let data: [String: Any] = [
            "routine": try! JSONEncoder().encode(routine),
            "feedback": feedback
        ]
        return await post(path: "/api/choreograph/refine", body: data)
    }
    
    /// 校验编舞合法性（不调 LLM）
    func validate(_ routine: DanceRoutine) async -> ChoreographResponse {
        return await post(path: "/api/choreograph/validate", body: routine)
    }
    
    // MARK: - 内部
    
    struct PromptResponse: Codable {
        let ok: Bool
        var prompt: String? = nil
        var backend: String? = nil
    }
    
    /// 查看当前 System Prompt
    func getPrompt() async -> PromptResponse {
        guard let url = URL(string: "\(baseURL)/api/choreograph/prompt") else {
            return PromptResponse(ok: false)
        }
        do {
            let (data, _) = try await session.data(from: url)
            return try decoder.decode(PromptResponse.self, from: data)
        } catch {
            return PromptResponse(ok: false)
        }
    }
    
    private func post(path: String, body: Encodable) async -> ChoreographResponse {
        guard let url = URL(string: "\(baseURL)\(path)") else {
            return ChoreographResponse(ok: false, error: "无效 URL")
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        do {
            request.httpBody = try encoder.encode(AnyEncodable(body))
        } catch {
            return ChoreographResponse(ok: false, error: "编码错误: \(error.localizedDescription)")
        }
        
        do {
            let (data, response) = try await session.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                return ChoreographResponse(ok: false, error: "HTTP 错误")
            }
            return try decoder.decode(ChoreographResponse.self, from: data)
        } catch {
            return ChoreographResponse(ok: false, error: "网络错误: \(error.localizedDescription)")
        }
    }
    
    private func post(path: String, body: [String: Any]) async -> ChoreographResponse {
        guard let url = URL(string: "\(baseURL)\(path)") else {
            return ChoreographResponse(ok: false, error: "无效 URL")
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        do {
            let (data, response) = try await session.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                return ChoreographResponse(ok: false, error: "HTTP 错误")
            }
            return try decoder.decode(ChoreographResponse.self, from: data)
        } catch {
            return ChoreographResponse(ok: false, error: "网络错误: \(error.localizedDescription)")
        }
    }
}

// ── AnyEncodable 包装器 ─────────────────────

struct AnyEncodable: Encodable {
    let value: Encodable
    
    init(_ value: Encodable) {
        self.value = value
    }
    
    func encode(to encoder: Encoder) throws {
        try value.encode(to: encoder)
    }
}
